contracts/hydro/src/contract.rs

```diff
@@ -676,135 +675,186 @@ fn vote(
+        // Get any existing vote for this sender and reverse it- this may be a vote for a different proposal (if they are switching their vote),
+        // or it may be a vote for the same proposal (if they have increased their power by locking more and want to update their vote).
+        // TODO: this could be made more gas-efficient by using a separate path with fewer writes if the vote is for the same proposal
+        let vote = VOTE_MAP.load(
             deps.storage,
-            (round_id, tranche_id, vote.prop_id),
-            &proposal,
-        )?;
+            ((round_id, tranche_id), info.sender.clone(), lock_id),
+        );
+        if let Ok(vote) = vote {
+            // Load the proposal in the vote
+            let mut proposal =
+                PROPOSAL_MAP.load(deps.storage, (round_id, tranche_id, vote.prop_id))?;
 
-        // Add proposal's new power in PROPS_BY_SCORE
-        if proposal.power > Uint128::zero() {
-            PROPS_BY_SCORE.save(
+            // Remove proposal's old power in PROPS_BY_SCORE
+            PROPS_BY_SCORE.remove(
                 deps.storage,
                 (
                     (round_id, proposal.tranche_id),
                     proposal.power.into(),
                     vote.prop_id,
                 ),
-                &vote.prop_id,
+            );
+
+            remove_validator_shares_from_proposal(
+                deps.storage,
+                round_id,
+                vote.prop_id,
+                vote.time_weighted_shares.0,
+                vote.time_weighted_shares.1,
             )?;
-        }
 
-        // Delete vote
-        VOTE_MAP.remove(deps.storage, (round_id, tranche_id, info.sender.clone()));
+            // save the new power into the proposal
+            let total_power = get_total_power_for_proposal(deps.as_ref().storage, vote.prop_id)?;
+            proposal.power = total_power.to_uint_ceil(); // TODO: decide whether we need to round or represent as decimals
 
-        response = response.add_attribute("old_proposal_id", vote.prop_id.to_string());
+            // Save the proposal
+            PROPOSAL_MAP.save(
+                deps.storage,
+                (round_id, tranche_id, vote.prop_id),
+                &proposal,
+            )?;
+
+            // Add proposal's new power in PROPS_BY_SCORE
+            if proposal.power > Uint128::zero() {
+                PROPS_BY_SCORE.save(
+                    deps.storage,
+                    (
+                        (round_id, proposal.tranche_id),
+                        proposal.power.into(),
+                        vote.prop_id,
+                    ),
+                    &vote.prop_id,
+                )?;
+            }
+
+            // Delete vote
+            VOTE_MAP.remove(
+                deps.storage,
+                ((round_id, tranche_id), info.sender.clone(), lock_id),
+            );
+
+            response = response.add_attribute("old_proposal_id", vote.prop_id.to_string());
```

Comment:
this should be refined to include the lock_id as part of the response, right?

Reply:
Added in https://github.com/informalsystems/hydro/pull/157/commits/f3f442ab6582ac8acd2ea812322edc73a28e7d10